### ISSUE TITLE
Incorporate Sondes DA into JEDI for HAFS

### ISSUE THREAD

**Issue by JingCheng-NOAA** [NOAA-EMC/HDASApp#16]
As the single ob DA is done within JEDI for HAFS, this issue is opened for more sounding obs DA in JEDI and comparison with GSI.
This issue will record the process for IODA file generation, JEDI yaml file for upper air DA, validation of Hx.

**Comment by JingCheng-NOAA** [comment]
## Start with the `airTemperature` with otype=120. Using the 2024 02L Beryl test case.
First split `ADPUPA` from gfs `prepbufr` file and edit the `gsiparm.anl` to take in OBS `t` only.
Then convert the GSI diag file into IODA NC file to provide observations for JEDI analysis.
The conversion produced two observation files:
`sondes_tsen_obs_2024063012.nc4`
`sondes_tv_obs_2024063012.nc4`

Run JEDI analysis with observation `tsen`, using `VertInterp`.
```
       obs operator:
         name: VertInterp
         variables:
         - name: airTemperature
       obs error:
         covariance model: diagonal
```
![fv3jedi_vs_gsi_validation_120_airTemperature](https://github.com/user-attachments/assets/8859b4ae-fd0b-4964-a619-bc37f1a11848)
![fv3jedi_vs_gsi_OMB_120_t](https://github.com/user-attachments/assets/b6314d7e-4ea2-4298-9b72-3fc60975c5af)

The hofx of the two are highly agree to each other. However the observation error in JEDI are smaller comparing with GSI

**Comment by JingCheng-NOAA** [comment]
Tried to inflated the error with ` ObsFunction/ObsErrorFactorConventional` and/or `ObsFunction/ObsErrorFactorPressureCheck`, but none of them can correct the error effectively.

**Comment by JingCheng-NOAA** [comment]
I've assigned the error as what HAFS GSI does with `errtable`, but the results of final ob error still doesn't look close to GSI.
```
           action:
             name: assign error
             error function:
               name: ObsFunction/ObsErrorModelStepwiseLinear
               options:
                 xvar:
                   name: MetaData/pressure
                 xvals: [110000, 105000, 100000, 95000, 90000, 85000, 80000, 75000, 70000, 65000, 60000, 55000, 50000, 45000, 40000, 35000, 30000, 25000, 20000, 15000, 10000, 7500, 5000, 4000, 3000, 2000, 1000, 500, 400, 300, 200, 100, 0]
                 errors: [1.2696, 1.3282, 1.3932, 1.439, 1.4354, 1.3669, 1.2552, 1.1362, 1.0397, 0.98016, 0.94757, 0.92875, 0.91714, 0.91625, 0.93506, 0.98447, 1.0699, 1.1816, 1.2862, 1.3542, 1.3831, 1.3832, 1.3534, 1.3109, 1.3018, 1.3414, 1.4017, 1.4471, 1.4744, 1.4892, 1.4964, 1.4988, 1.4962]

```

**Comment by JingCheng-NOAA** [comment]
Adding observation inflation using `ObsFunction/ObsErrorFactorPressureCheck`, but the final observation error of JEDI is unchanged as not use. Guess need to find a better way to inflate the observation error.
added section:
```
         - filter: Perform Action
           filter variables:
           - name: airTemperature
           where:
           - variable: ObsType/airTemperature
             is_in: 120
           action:
             name: inflate error
             inflation variable:
               name: ObsFunction/ObsErrorFactorPressureCheck
               options:
                 variable: airTemperature
                 inflation factor: 8.0
                 #geovar_sfc_geomz: surface_geometric_height
           defer to post: true

```

**Comment by JingCheng-NOAA** [comment]
Test on specificHumidity
![fv3jedi_vs_gsi_validation_120_specificHumidity](https://github.com/user-attachments/assets/76dafd43-376b-45d5-90aa-8715c0d503a6)

**Comment by JingCheng-NOAA** [comment]
Test with uv. Here for the wind observations, introducing hofx scaling into the observation operator.
Use GSI geovals will get identical hofx.
![GSIgeovals_hofx_220_windEastward](https://github.com/user-attachments/assets/63066c1f-23aa-4024-aac9-7e0e69a05764)

if not use GSI geovals, then there are slight difference in JEDI and GSI hofx. The hofx scaling also did some correction here.
![jedi_gsi_hofx_220_windEastward](https://github.com/user-attachments/assets/a9c34324-62bc-4206-83f8-78f304c922c9)


HOFX after scaling - HOFX no scaling
![fv3jedi_hofxscaling_vert_validation_220_windEastward](https://github.com/user-attachments/assets/af847caf-b76e-4a8d-8a79-3a082c28a6e4)

**Comment by JingCheng-NOAA** [comment]
Rerun the GSI with `grid_ratio_fv3_regional=1`, but still find the difference in hofx is large.
Plot out the difference of hofx and found the large mismatch on the top of model level.

![hofxhist_windEastward](https://github.com/user-attachments/assets/9a26d955-6b71-4bf7-9b35-0f5c264189d6)
![hofxdiff_vs_pressure_windEastward](https://github.com/user-attachments/assets/08715914-5980-4e48-b5a0-357195d31c96)
![hofxhist_windNorthward](https://github.com/user-attachments/assets/ddbe371f-22b7-472f-bfc9-f6a73cb9f4e0)
![hofxdiff_vs_pressure_windNorthward](https://github.com/user-attachments/assets/f004db07-a21a-481f-ac62-6e60d4fe9371)

**Comment by JingCheng-NOAA** [comment]
Though the IODA NC file is created from GSI diag file, but it tends out several points are out of JEDI domain boundary. Using the RRFS domain filter tool to further filter out points outside or near the boundary.
![domain_check_FV3](https://github.com/user-attachments/assets/5f44c1b0-ad0a-4dc9-b8fd-7d561dc84d0f)

**Comment by JingCheng-NOAA** [comment]
After applying the domain check, the hofx difference between jedi and gsi are smaller. But the large difference near top of the model remains.
![domainfilter_fv3jedi_vs_gsi_validation_220_windEastward](https://github.com/user-attachments/assets/31cab8bc-ecea-435d-a049-59b193c057f1)
![domainfilter_fv3jedi_vs_gsi_validation_220_windNorthward](https://github.com/user-attachments/assets/0c634a62-04ca-4f3f-ba70-d7af8d5c0ed0)
![domainfilter_hofxhist_windEastward](https://github.com/user-attachments/assets/7fd54cb9-79ef-46c3-93ae-0d790c21ac1e)
![domainfilter_hofxdiff_vs_pressure_windEastward](https://github.com/user-attachments/assets/75e266ed-81f1-4f32-bd96-5b5fbe4210a4)
![domainfilter_hofxhist_windNorthward](https://github.com/user-attachments/assets/10b086e9-da27-47eb-87f3-a218c2d63601)
![domainfilter_hofxdiff_vs_pressure_windNorthward](https://github.com/user-attachments/assets/2e20a21e-a831-4201-ad6c-32c60c088741)

**Comment by JingCheng-NOAA** [comment]
Select the point where has max hofx difference, the location is lat=4.75 lon=285.80
FROM JEDI DIAG:
obs index 1431, location: 4.75 285.80,
pressure:  **8419.999694824219**,
obsvalue:-10.199999809265137
jhofx: -12.158164978027344, _**ghofx: -13.30904769897461**_

FROM GSI DIAG:
matched obs index 587
location: 4.754079818725586 285.79833984375,
**_eastward_wind -13.183977127075195_**
pressure level: **8431.2216796875**
u_obs:-10.199999809265137

FROM ncdiag file converted IODA nc:
 location: 4.754079818725586 285.79833984375
 obs value -10.199999809265137
FROM bufr2ioda converted IODA nc:
 location: 4.754079818725586 -74.20166015625
 obs value -10.199999809265137

**Comment by JingCheng-NOAA** [comment]
Check the gsi and jedi geoval files to see the "air_pressure" distribution difference. GSI geofiles contains lat and lon information, so it's easier to locate the obs location that located near "4.75 285.80".
For JEDI, all the files do not contain any lat/lon information, the only way to locate is using the index number the same as we found in GSI geofiles, the results are quite different too

![geoval_airpressure](https://github.com/user-attachments/assets/f198766a-9bb6-4c8e-9fac-a0390ed3a930)

From GSI
surface pressure 75540.2734375
air pressure [  325.       600.       957.5685  1440.7946  2017.909   2646.323
  3325.437   4054.6511  4833.365   5660.979   6536.893   7460.5073
  8431.222   9448.436  10494.844  11540.1    12575.047  13607.0205
 14641.574  15682.952  16734.43   17798.547  18877.28   19972.168
 21084.408  22214.926  23364.422  24533.428  25722.33   26931.395
 28160.795  29410.621  30680.898  31971.592  33282.613  34613.848
 35965.13   37336.27   38727.06   40137.258  41566.61   43013.66
 44474.168  45942.13   47411.387  48875.785  50329.164  51765.37
 53178.25   54561.9    55910.97   57220.7    58486.99   59706.445
 60876.363  61994.676  63059.98   64071.48   65028.94   65932.64
 66783.3    67581.99   68330.1    69029.31   69681.47   70288.586
 70852.86   71376.42   71861.51   72310.33   72725.13   73108.086
 73461.21   73786.61   74086.26   74361.95   74615.46   74848.445
 75062.484  75259.05   75446.484 ]

For JEDI,
surface pressure [96645.22]
air pressure [  319.10495   595.48126   952.16327  1434.4716   2012.5248   2641.5
  3320.988    4050.47     4829.386    5657.159    6533.2026   7456.9243
  8427.73     9445.024   10505.213   11602.284   12734.048   13901.352
 15104.716   16344.424   17620.58    18933.156   20282.023   21666.967
 23087.717   24543.947   26035.299   27561.373   29121.754   30715.994
 32343.637   34004.207   35697.22    37422.17    39178.56    40965.87
 42783.586   44631.184   46508.13    48413.887   50347.92    52308.094
 54288.496   56280.83    58276.555   60267.156   62244.094   64198.84
 66122.875   68008.016   69846.836   71632.73    73359.984   75023.88
 76620.63    78147.336   79602.      80983.48    82291.38    83526.05
 84688.414   85779.914   86802.414   87758.18    88649.71    89479.75
 90251.266   90967.18    91630.51    92244.29    92811.58    93335.336
 93818.32    94263.39    94673.27    95050.41    95397.19    95715.91
 96008.71    96277.62    96525.76   ]

**Comment by JingCheng-NOAA** [comment]
The index match method may not correct. So check the 1st point in both GSI and JEDI GeoVals.
From GSI for idx 0
    surface pressure 101772.7421875
From JEDI for idx 0
    surface pressure 101786.336
![geoval_airpressure_0point](https://github.com/user-attachments/assets/c1d8290e-85b6-4da4-91b5-542b1a8c1cc2)

**Comment by JingCheng-NOAA** [comment]
Found a bug in finding the matched point index in JEDI GEOVaLs.
And @delippi mentioned about the difference in air_pressure calculation recipes in GSI and JEDI [https://github.com/NOAA-EMC/RDASApp/issues/54#issuecomment-2150339798](url).
So modified the code to use the same middle point of pressure level and it did change the first 15 levels of air_pressure values. 
![geoval_airpressure](https://github.com/user-attachments/assets/ecbcdc44-4e49-42ef-81e4-7da8a39dd1de)
![geoval_GEOVAL_uwind](https://github.com/user-attachments/assets/9f237f64-0409-4cde-a7c6-7ca637efd043)
![geoval_GEOVAL_vwind](https://github.com/user-attachments/assets/8c66621a-c8d2-466d-98da-e8d48834e17f)

From gsi geoval file nc:
 idx 587
 location 4.754079818725586 285.79833984375
 surface pressure 75540.2734375
air pressure [  **325.       600.       957.5685  1440.7946  2017.909   2646.323
  3325.437   4054.6511  4833.365   5660.979   6536.893   7460.5073
  8431.222   9448.436  10494.844**  11540.1    12575.047  13607.0205
 14641.574  15682.952  16734.43   17798.547  18877.28   19972.168
 21084.408  22214.926  23364.422  24533.428  25722.33   26931.395
 28160.795  29410.621  30680.898  31971.592  33282.613  34613.848
 35965.13   37336.27   38727.06   40137.258  41566.61   43013.66
 44474.168  45942.13   47411.387  48875.785  50329.164  51765.37
 53178.25   54561.9    55910.97   57220.7    58486.99   59706.445
 60876.363  61994.676  63059.98   64071.48   65028.94   65932.64
 66783.3    67581.99   68330.1    69029.31   69681.47   70288.586
 70852.86   71376.42   71861.51   72310.33   72725.13   73108.086
 73461.21   73786.61   74086.26   74361.95   74615.46   74848.445
 75062.484  75259.05   75446.484 ]
From jedi geoval file nc:
 idx 354
 location 28.25101089477539 279.1451416015625
 surface pressure [75602.24]
air pressure [  **325.       600.       957.5685  1440.7946  2017.909   2646.323
  3325.437   4054.6511  4833.365   5660.979   6536.893   7460.5073
  8431.222   9448.436  10494.884**  11540.291  12575.522  13607.895
 14642.944  15684.905  16737.045  17801.893  18881.422  19977.166
 21090.314  22221.791  23372.297  24542.355  25732.355  26942.557
 28173.13   29424.17   30695.695  31987.672  33300.01   34632.594
 35985.254  37357.81   38750.035  40161.7    41592.543  43041.113
 44503.16   45972.67   47443.49   48909.44   50364.37   51802.113
 53216.504  54601.64   55952.16   57263.297  58530.95   59751.727
 60922.906  62042.426  63108.887  64121.484  65079.98   65984.66
 66836.234  67635.79   68384.71   69084.69   69737.55   70345.33
 70910.21   71434.34   71919.95   72369.26   72784.516  73167.88
 73521.4    73847.15   74147.12   74423.12   74676.9    74910.14
 75124.414  75321.195  75508.805 ]

At least we know the difference in hofx shouldn't be coming from air_pressure.
![hofxdiff_vs_pressure_windEastward_modifiedReceipe](https://github.com/user-attachments/assets/9a96c26b-b62d-4f84-94bd-4c47245d385f)

**Comment by delippi** [comment]
@JingCheng-NOAA, Thanks for testing this. It looks like you are using GSI-IODA but not using GSI-geovals and just letting JEDI compute its own geovals, is that correct? Did you say that you get identical results when using the GSI-geovals? Can you add the the same pressure vs hofx difference plot for when using GSI-geovals?

**Comment by JingCheng-NOAA** [comment]
> @JingCheng-NOAA, Thanks for testing this. It looks like you are using GSI-IODA but not using GSI-geovals and just letting JEDI compute its own geovals, is that correct? Did you say that you get identical results when using the GSI-geovals? Can you add the the same pressure vs hofx difference plot for when using GSI-geovals?

Yes the results of using same GSI GeoVaLs are showed in [https://github.com/NOAA-EMC/HDASApp/issues/16#issuecomment-2372193384](url)
![GSIgeovals_hofx_220_windEastward](https://github.com/user-attachments/assets/e406eae1-a6f3-4d77-a8c2-eb94ad8f25af)


And if plot the difference in vertical distribution:
![GSIGeoVals_hofxdiff_vs_pressure_windEastward](https://github.com/user-attachments/assets/20526826-5e46-40f7-bc02-f5c0beb08d25)

**Comment by JingCheng-NOAA** [comment]
But the large difference in GPH remains large.
![geoval_gph](https://github.com/user-attachments/assets/ed6542ed-5cce-4a8a-a7ed-26ab57f987d7)

**Comment by delippi** [comment]
@JingCheng-NOAA, interesting. I would expect that if you get identical results with GSI-geovals then you would have identical results at levels 1-15 and still have differences elsewhere. But you might be using gph as vertical coordinate or something else instead of pressure. Check what is used in GSI.

**Comment by JingCheng-NOAA** [comment]
> @JingCheng-NOAA, interesting. I would expect that if you get identical results with GSI-geovals then you would have identical results at levels 1-15 and still have differences elsewhere. But you might be using gph as vertical coordinate or something else instead of pressure. Check what is used in GSI.

I checked the setupw.f90 in GSI and for this type of observations (otype=220), the GSI is using pressure levels.

**Comment by JingCheng-NOAA** [comment]
An update was made in comparison GSI test, block the otype 232 and only keep the otype 220.
Without domain filter, the results are updated here
![fv3jedi_vs_gsi_validation_220_windEastward](https://github.com/user-attachments/assets/8ca4b3de-c4a4-4c80-a433-36b3a637dda2)
![fv3jedi_vs_gsi_validation_220_windNorthward](https://github.com/user-attachments/assets/2730ee09-9ab3-474f-9313-5174035f7a09)

**Comment by JingCheng-NOAA** [comment]
![fv3jedi_vs_gsi_validation_120_specificHumidity](https://github.com/user-attachments/assets/30341d85-8a2e-43c9-a71d-4b0d1634c1f6)

**Comment by JingCheng-NOAA** [comment]
![fv3jedi_vs_gsi_validation_120_stationPressure](https://github.com/user-attachments/assets/78e0ce5c-e264-43dc-8072-1d05b558453d)

**Comment by JingCheng-NOAA** [comment]
![fv3jedi_vs_gsi_validation_120_airTemperature](https://github.com/user-attachments/assets/5e3f1616-af69-405c-bf42-274946289cca)

**Comment by JingCheng-NOAA** [comment]
![fv3jedi_vs_gsi_validation_120_virtualTemperature](https://github.com/user-attachments/assets/6b47e1dd-7a8b-451c-bb36-7dc2eb1ca2a0)

**Comment by JingCheng-NOAA** [comment]
I've noticed this discussion in GDASApp, which compared the difference in Cube and Gaussian grids. This may explain our issue found in the difference of geovals in GSI and JEDI.
[https://github.com/NOAA-EMC/GDASApp/issues/750#issuecomment-2079617045](url)
ps, similar difference in sounding data was found in GDAS test too 
[https://github.com/NOAA-EMC/GDASApp/issues/750#issuecomment-2077646953](url)

**Comment by JingCheng-NOAA** [comment]
Move to phase 2 for sounding data. 
Start with prepareing the IODA nc using the bufr2ioda in GDASApp for wind observations (otype=220). 
Meanwhile, block other observation types of data and only keep otype=220 in GSI parm file.
Observation count difference in GSI and JEDI diag file. Need to find out which data were rejected by GSI but accepted by JEDI
GSI: 1456
JEDI: 1614

**Comment by JingCheng-NOAA** [comment]
Modified the bufr2ioda python script to add in MetaData/timeOffset. This variable will be used to QC the observation that fall out of time window

**Comment by JingCheng-NOAA** [comment]
Regenerated IODA nc file with updated bufr2ioda yaml file.
Added in the on-line domain filter with updated HDASApp submodules.
The number used by JEDI is reduced a lot comparing with GSI.
![obsnumber](https://github.com/user-attachments/assets/1be09bf4-0a9a-4753-baa8-e30916fbd55e)

Here are some statistics:
*JEDI without any obs filter:
- Number used in analysis: 1150
- Missing value prevent use of observation: 30510
- hofx computation failed: 37399

*JEDI with obs filter applied: 
- Number used in analysis: 1138
- Missing value prevent use of observation: 30510
- Observation value out of bounds: 35327
- Observation not within domain: 878
- Observation removed due to thinning: 1206

**Comment by JingCheng-NOAA** [comment]
The observation distribution plot illustrated where the missing observations in JEDI are.
![image](https://github.com/user-attachments/assets/b213b572-c464-4576-876f-609803d323eb)

**Comment by JingCheng-NOAA** [comment]
Updated IODA yaml file with correct time and recreated observation file shows correct observation numbers that similar to GSI.
Also fixed a bug in python script to correctly calculate the wind increment of GSI.
![diag_wind_220_hofxobs_on_domain](https://github.com/user-attachments/assets/91429cbc-7549-45bd-932a-dd405580e659)

**Comment by JingCheng-NOAA** [comment]
Check for airTemperature and specificHumidity
![diag_t_120_hofxobs_on_domain](https://github.com/user-attachments/assets/1928c479-dc77-4d18-b9bb-50c85209fb73)
![diag_q_120_hofxobs_on_domain](https://github.com/user-attachments/assets/11470d8e-f932-40ea-a1bc-3df413084c87)

**Comment by JingCheng-NOAA** [comment]
Thanks to @delippi 's plotting scripts, I've checked the stats after QC for each variable. Some of the observation error still has difference.
![windstats](https://github.com/user-attachments/assets/d4578e38-8388-4ebb-bead-650cf513d3ab)
![airTemperatureStats](https://github.com/user-attachments/assets/10ecf2eb-4dbe-4950-9088-e6e273bc5f37)
![specificHumidityStats](https://github.com/user-attachments/assets/95c8d767-0d35-4525-9b2b-bc7dcb75f06e)

**Comment by JingCheng-NOAA** [comment]
An update was made in Wind yaml file, including adjusting the gross error and a bug fix in `ObsErrorFactorConventional`. Meanwhile, rerun the GSI with assimilate wind 220 only (previously was assimilating every type of observations).

![image](https://github.com/user-attachments/assets/5c08064d-5a18-4d7b-8563-92057f310295)

**Comment by JingCheng-NOAA** [comment]
For specific Humidity, adjusted the gross error min/max in JEDI yaml; And fixed a bug in the initial error assignment in JEDI yaml (the value grabbed from GSI errtable should be divided by 10 for q)
![image](https://github.com/user-attachments/assets/51cfe276-3261-40da-8f4d-b2434964113e)

**Comment by JingCheng-NOAA** [comment]
The observation error pattern mismatch in airTemperature. 
Set up single observation test and compare the error.
![image](https://github.com/user-attachments/assets/c626995d-b8ac-44c7-9f32-7f6e61f93351)

**Comment by JingCheng-NOAA** [comment]
In GSI read_prepbufr.f90, the error is inflated by 1.2, which leads to
```
final error = input_error*1.2 = 1.369138*1.2 = 1.642966
```
I am wondering which function in JEDI can do similar inflation.

**Comment by delippi** [comment]
@JingCheng-NOAA, yep this should be fairly easy and probably a few ways to do it. One of the easiest ways would probably be to use the `inflate error` action in any filter that you could use the `reject` action and just make sure it selects all observations. I didn't test the following configuration but it might look something like this. I can test it when Hera comes back.
```
- filter: Background Check
  filter variables:
  - name: airTemperature
action:
    name: inflate error
    inflation: 1.2
```

**Comment by JingCheng-NOAA** [comment]
Thank you Donnie @delippi , this is very helpful. I will test it too once Hera comes back.

**Comment by delippi** [comment]
@JingCheng-NOAA, if you hadn't already figured out what was wrong with my previous try, this works. I was close.
```
         - filter: Perform Action
           filter variables:
           - name: airTemperature
           action:
             name: inflate error
             inflation factor: 1.2
```

**Comment by JingCheng-NOAA** [comment]
Thanks, @delippi . I've tried this one and it works for my single observation test. The observation error in JEDI now looks the same as in GSI single ob exp.

**Comment by JingCheng-NOAA** [comment]
After apply the inflation fix to the phase 2 test, the results look improved.
Bottom left, previous observation error in airTemperature; Bottom right, added the inflation in observation error.
![image](https://github.com/user-attachments/assets/e51fa801-3827-430b-b9b8-31703954ec8a)

**Comment by delippi** [comment]
@JingCheng-NOAA, for the 1.2 inflation, it looks like this is done for all conventional obs and for when QualityMarker==3 or 7. Is that correct? I think I should add this to my configurations.

**Comment by JingCheng-NOAA** [comment]
@delippi, according to read_prepbufr.f90, this inflation only happen to the Temperature observation on pressure level 100hPa or above; and wind observation on pressure level 50hPa or above. I think the QM is irrelevant here. So based on what you taught me in previous comment, I added following part right after the initial error assignment,
```
 ## error inflation if observation pressure < 100 hPa (read_prepbufr.f90)
         - filter: Perform Action
           filter variables:
           - name: airTemperature
             where:
              - variable: MetaData/pressure
             maxvalue: 10000.
           action:
             name: inflate error
             inflation factor: 1.2
```

**Comment by delippi** [comment]
Actually, I think we're both only half right. I see the following in read_prepbufr for temperature. The obs can get inflated twice by 1.2.
```
                 if (inflate_error) toe=toe*r1_2
                 if(ppb < r100)toe=toe*r1_2
```

inflate error comes from
```
              inflate_error=.false.
              if (qm==3 .or. qm==7) inflate_error=.true.
```

**Comment by JingCheng-NOAA** [comment]
Yes I think you are right. My case doesn't have qm 3 or qm 7 that's why I thought the inflate_error is disabled. But this part should be added in too.

**Comment by JingCheng-NOAA** [comment]
Updated the yaml file to added in the qm 3 or qm 7 inflation.

![Image](https://github.com/user-attachments/assets/5091c269-70dc-440f-9614-ffaf8b67fed9)

**Comment by JingCheng-NOAA** [comment]
The remain missing observation error inflation is related to the duplication check that designed in setupt.f90. The corresponding obsFunction in JEDI is `ObsErrorFactorDuplicateCheck`, which was wrote and tested with GSI diag converted IODA file. The variable `MetaData/PreUseFlag` does not exist in current bufr2ioda converted NC file. Donnie suggested to either add this variable into IODA nc file through an external python script or modify the c++ code to change the usage of this variable. 
I tried to add the PreUseFlag with a preset value 0 into the IODA nc file, which turns out to modify a lot of observation errors, which shouldn't be correct, as in this testcase, the GSI results show that no duplicated observation. It also will be questionable to add this arbitration value to PreUseFlag.
Thus, I modified the c++ code  to replace the variable `PreUseFlag` with `testQCflag`. The test with this change run through and the observation error pattern stay unchanged. Since the dataset contains no duplicated observation, the results look good. 
Meanwhile, an issue was open in JCSDA/ufo [https://github.com/JCSDA-internal/ufo/issues/3593](url) to report the missing variable issue in the obsfunction.

So far, besides stationPressure (requires further changes in IODA converter), the phase 2 validation of sonde yamls is considered completed.

**Comment by JingCheng-NOAA** [comment]
Revisit this issue with updates on the duplication check. This function `ObsErrorFactorDuplicateCheck` is functional again by assigning value into `PreUseFlag`. Validation has been done on a smaller domain size, which ensure the same coverage of area on both GSI and JEDI.

![Image](https://github.com/user-attachments/assets/b32674fc-af2b-439b-921e-0cd9b8f87cf5)

![Image](https://github.com/user-attachments/assets/5c160b5e-6f3e-412a-b147-b09a3509bf68)

![Image](https://github.com/user-attachments/assets/a4804858-fc3d-4146-b4d4-9a3d26d81fe0)

**Comment by JingCheng-NOAA** [comment]
Another update is the IODA process is able to separate the Tv and Tsen from Prepbufr following GSI procedure. Thus the validation of airTemperature will be for the Tsen.

![Image](https://github.com/user-attachments/assets/7390bf31-ae83-47d8-944f-d434cd944e9a)
